crates/stages/src/stages/tx_lookup.rs

```diff
@@ -0,0 +1,295 @@
+use crate::{
+    db::Transaction, exec_or_return, ExecAction, ExecInput, ExecOutput, Stage, StageError, StageId,
+    UnwindInput, UnwindOutput,
+};
+use reth_db::{
+    cursor::{DbCursorRO, DbCursorRW},
+    database::Database,
+    tables,
+    transaction::{DbTx, DbTxMut},
+};
+use tracing::*;
+
+const TRANSACTION_LOOKUP: StageId = StageId("TransactionLookup");
+
+/// The transaction lookup stage.
+///
+/// This stage walks over the bodies table, and sets the transaction hash of each transaction in a
+/// block to the corresponding `TransitionId` at each block. This is written to the
+/// [`tables::TxHashNumber`] This is used for looking up changesets via the transaction hash.
+#[derive(Debug)]
+pub struct TransactionLookupStage {
+    /// The number of table entries to commit at once
+    pub commit_threshold: u64,
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for TransactionLookupStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        TRANSACTION_LOOKUP
+    }
+
+    /// Write total difficulty entries
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let ((start_block, end_block), capped) =
+            exec_or_return!(input, self.commit_threshold, "sync::stages::transaction_lookup");
+
+        debug!(target: "sync::stages::transaction_lookup", start_block, end_block, "Commencing sync");
+
+        let mut cursor_bodies = tx.cursor_read::<tables::BlockBodies>()?;
+        let mut tx_cursor = tx.cursor_write::<tables::Transactions>()?;
+        let start_key = tx.get_block_numhash(start_block)?;
+
+        // Walk over block bodies within a specified range.
+        let bodies = cursor_bodies
+            .walk(start_key)?
+            .take_while(|entry| {
+                entry
+                    .as_ref()
+                    .map(|(block_num_hash, _)| block_num_hash.number() <= end_block)
+                    .unwrap_or_default()
+            })
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
is this collect required because of the subsequent `tx_cursor.walk` over transactions?
in any case, looks like we only need the `start_tx_id` and `tx_count`